Add migration guide (#23482)
microsoft/onnxruntime: docs/genai/howto/migrate.md

---
title: Migrate
description: Learn how to migrate from one version of ONNX Runtime generate() API when there are breaking API changes
has_children: false
parent: How to
grand_parent: Generate API (Preview)
nav_order: 5
---

> # Migrate ONNX Runtime generateA() API from 0.5.2 to 0.6.0

Reviewer: Should rewind be mentioned somewhere in this guide?

Author: Can you give me a succinct description of the function and purpose of the rewind API?

Reviewer: rewind_to(int new_length) rewinds the generator back to when it had new_length tokens and can be used, for example, to rewind back to the system prompt, to reset the generator (i want to add a proper reset api though), to rewind back to a previous point in the conversation, to remove tokens we didn't like that the generator generated etc.

Learn how to migrate from ONNX Runtime generate() version 0.5.2 to version 0.6.0. 

Version 0.6.0 adds support for "chat mode", also known as _continuation_, _continous decoding_, and _interactive decoding_. The introduction of chat mode necessitated a change to the API, which breaks the previous API.

> In summary, the new API adds support for `AppendTokens`, which allows turn taking in the conversation. Previously, there was a simple API to `SetInputs`.

Reviewer: ```suggestion
In summary, the new API adds an `AppendTokens` function to the generator, which allows for multi-turn conversations. Previously, input was set in `GeneratorParams` prior to the creation of the generator.
```

> Calling `AddTokens` outside of the loop also adds support for system prompt caching.

Reviewer: ```suggestion
Calling `AppendTokens` outside of the conversation loop can be used to implement system prompt caching.
```

> Note: chat mode and system prompt caching is only supported when running on CPU, NVIDIA GPUs with the CUDA EP, and all GPUs with the Web GPU native EP. It is not supported on NPU or GPUs running with the DirecML EP. For Q&A mode, the migrations described below *are* required.

Reviewer: ```suggestion
Note: chat mode and system prompt caching are only supported for batch size 1. Furthermore, they are currently supported on CPU, NVIDIA GPUs with the CUDA EP, and all GPUs with the Web GPU native EP. They are not supported on NPU or GPUs running with the DirecML EP. For Q&A mode, the migrations described below *are* required.
```

## Python

### Migrate Python question and answer (single turn) code to 0.6.0

1. Replace calls to `params.input_ids = input_tokens` with `generator.append_tokens(input_tokens)` after the generator object has been created.
2. Remove calls to `generator.compute_logits()`

> 

Reviewer: For Q&A _without_ multi-turn conversations you would also rewind (or delete the generator) after the generation loop

### Add system prompt caching

1. Create and tokenize the system prompt and call `generator.append_tokens(system_tokens)`. This call can be done before the user is asked for their prompt.

   ```python
   system_tokens = tokenizer.encode(system_prompt)
   generator.append_tokens(system_tokens)
   ```

### Add chat mode

1. Create a loop in your application, and call `generator.append_tokens(prompt)` every time the user provides new input:
   
   ```python
   while True:
       user_input = input("Input: ")
       input_tokens = tokenizer.encode(user_input)
       generator.append_tokens(input_tokens)

       while not generator.is_done():
           generator.generate_next_token()

           new_token = generator.get_next_tokens()[0]
           print(tokenizer_stream.decode(new_token), end='', flush=True)
        except KeyboardInterrupt:
        print()
    ```

## C/C++ 

### Migrate C/C++ question and answer (single turn) code to 0.6.0

> 1. Replace calls to `params->SetInputSequences(*sequences)` with `generator->AppendTokenSequences(*sequences)`

Reviewer: We also have the AppendTokens option but I guess that's usually used without a tokenizer

Author: What is the use case for not using a tokenizer? Is that for models other than language ones?

Reviewer: We use it to set tokens directly for test cases, not sure about other use cases
